source/llassetgen/source/DistanceTransform.cpp

```diff
@@ -1,300 +1,162 @@
-        png_destroy_read_struct(&png, &pngInfo, nullptr);
-        fclose(file);
-    }
-
-    template <typename PixelType>
-    void DistanceTransform::exportPngInternal(png_struct* png, OutputType blackDistance, OutputType whiteDistance) {
-        std::unique_ptr<PixelType[]> rowBuffer(new PixelType[width]);
-        for (DimensionType y = 0; y < height; ++y) {
-            for (DimensionType x = 0; x < width; ++x)
-                rowBuffer[x] = clamp((outputAt({x, y}) - blackDistance) / (whiteDistance - blackDistance), 0.0F, 1.0F) *
-                               std::numeric_limits<PixelType>::max();
-            png_write_row(png, reinterpret_cast<png_bytep>(rowBuffer.get()));
-        }
-    }
-
-    void DistanceTransform::exportPng(std::string path, OutputType blackDistance, OutputType whiteDistance,
-                                      DimensionType bitDepth) {
-        assert(width > 0 && height > 0 && output.get());
-        FILE* file = fopen(path.c_str(), "w");
-        assert(file);
-        png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
-        png_infop pngInfo = png_create_info_struct(png);
-        if (setjmp(png_jmpbuf(png)))
-            assert(false);
-        else {
-            png_init_io(png, file);
-            png_set_IHDR(png, pngInfo, width, height, bitDepth, PNG_COLOR_TYPE_GRAY, PNG_INTERLACE_NONE,
-                         PNG_COMPRESSION_TYPE_BASE, PNG_FILTER_TYPE_BASE);
-            png_write_info(png, pngInfo);
-            png_set_swap(png);
-            switch (bitDepth) {
-                case 8:
-                    exportPngInternal<unsigned char>(png, blackDistance, whiteDistance);
-                    break;
-                case 16:
-                    exportPngInternal<unsigned short>(png, blackDistance, whiteDistance);
-                    break;
-                default:
-                    assert(false);
-            }
-        }
-        png_write_end(png, nullptr);
-        png_destroy_write_struct(&png, &pngInfo);
-        fclose(file);
-    }
 
     DeadReckoning::PositionType& DeadReckoning::posAt(PositionType pos) {
-        assert(pos.x < width && pos.y < height && posBuffer.get());
-        return posBuffer[pos.y * width + pos.x];
+        assert(pos.x < input.get_width() && pos.y < input.get_height() && posBuffer.get());
+        return posBuffer[pos.y * input.get_width() + pos.x];
     }
 
     void DeadReckoning::transformAt(PositionType pos, PositionType target, OutputType distance) {
         target += pos;
-        if (outputAtClamped(target) + distance < outputAt(pos)) {
+        if (output.is_valid(target) && getPixel<OutputType>(target) + distance < getPixel<OutputType>(pos)) {
             posAt(pos) = target = posAt(target);
-            outputAt(pos) = std::sqrt(square(pos.x - target.x) + square(pos.y - target.y));
+            setPixel<OutputType>(pos, std::sqrt(square(pos.x - target.x) + square(pos.y - target.y)));
```

Comment:
I think glm [1] would be a valid dependency for this project.

[1] https://github.com/g-truc/glm

Reply:
I tried using alternatives like `std::hypot(x, y)` but it is not that easy because of the signed / unsigned types

Reply:
In llassetgen-rendering I use glm, too. So if we need glm as a main dependency, I would benefit from it, too.